File: Codigo Python/GestionDeCursos.py
import sys
from PyQt5.QtWidgets import QApplication, QWidget, QMainWindow, QLabel, QLineEdit, QPushButton, QVBoxLayout, QHBoxLayout, QMessageBox, QComboBox, QTextEdit, QStackedWidget, QDialog


class Profesor:
    def __init__(self, nombre, apellido):
        self.nombre = nombre
        self.apellido = apellido

    def getNombreCompleto(self):
        return f"{self.nombre} {self.apellido}"


class Curso:
    def __init__(self, nombre, profesor):
        self.nombre = nombre
        self.profesor = profesor
        self.estudiantes = []

    def agregarEstudiante(self, estudiante):
        self.estudiantes.append(estudiante)


class Estudiante:
    def __init__(self, nombre, apellido):
        self.nombre = nombre
        self.apellido = apellido

    def getNombreCompleto(self):
        return f"{self.nombre} {self.apellido}"


class GestionDeCursos(QMainWindow):
    def __init__(self):
        super().__init__()
        self.cursos = []
        self.profesores = []
        self.initUI()

    def initUI(self):
        self.setWindowTitle("Gestión de Cursos")
        self.setGeometry(100, 100, 600, 400)

        self.central_widget = QWidget()
        self.setCentralWidget(self.central_widget)

        self.stacked_widget = QStackedWidget()
        layout = QVBoxLayout(self.central_widget)
        layout.addWidget(self.stacked_widget)

        # Página principal
        self.pagina_principal = QWidget()
        self.stacked_widget.addWidget(self.pagina_principal)

        self.layout_principal = QVBoxLayout(self.pagina_principal)

        self.label_info = QLabel("Bienvenido a la Gestión de Cursos")
        self.layout_principal.addWidget(self.label_info)

        button_agregar_curso = QPushButton("Agregar Curso")
        button_agregar_curso.clicked.connect(self.mostrarVentanaAgregarCurso)
        self.layout_principal.addWidget(button_agregar_curso)

        button_agregar_profesor = QPushButton("Agregar Profesor")
        button_agregar_profesor.clicked.connect(self.mostrarVentanaAgregarProfesor)
        self.layout_principal.addWidget(button_agregar_profesor)

        button_agregar_estudiante = QPushButton("Agregar Estudiante")
        button_agregar_estudiante.clicked.connect(self.mostrarVentanaAgregarEstudiante)
        self.layout_principal.addWidget(button_agregar_estudiante)

        button_ver_horarios = QPushButton("Ver Horarios")
        button_ver_horarios.clicked.connect(self.mostrarVentanaHorarios)
        self.layout_principal.addWidget(button_ver_horarios)

        # Inicializamos las páginas de agregar curso, profesor, estudiante y horarios
        self.ventana_agregar_curso = VentanaAgregarCurso(self)
        self.stacked_widget.addWidget(self.ventana_agregar_curso)

        self.ventana_agregar_profesor = VentanaAgregarProfesor(self)
        self.stacked_widget.addWidget(self.ventana_agregar_profesor)

        self.ventana_agregar_estudiante = VentanaAgregarEstudiante(self)
        self.stacked_widget.addWidget(self.ventana_agregar_estudiante)

        self.ventana_horarios = VentanaHorarios(self)
        self.stacked_widget.addWidget(self.ventana_horarios)

    def mostrarVentanaAgregarCurso(self):
        self.stacked_widget.setCurrentWidget(self.ventana_agregar_curso)

    def mostrarVentanaAgregarProfesor(self):
        self.stacked_widget.setCurrentWidget(self.ventana_agregar_profesor)

    def mostrarVentanaAgregarEstudiante(self):
        self.stacked_widget.setCurrentWidget(self.ventana_agregar_estudiante)

    def mostrarVentanaHorarios(self):
        self.stacked_widget.setCurrentWidget(self.ventana_horarios)


class VentanaAgregarCurso(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.parent = parent
        self.initUI()

    def initUI(self):
        self.layout = QVBoxLayout()

        self.label_nombre_curso = QLabel("Nombre del Curso:")
        self.nombre_curso_field = QLineEdit()
        self.layout.addWidget(self.label_nombre_curso)
        self.layout.addWidget(self.nombre_curso_field)

        self.label_profesor = QLabel("Profesor:")
        self.profesor_combo = QComboBox()
        for profesor in self.parent.profesores:
            self.profesor_combo.addItem(profesor.getNombreCompleto())
        self.layout.addWidget(self.label_profesor)
        self.layout.addWidget(self.profesor_combo)

        self.button_agregar_curso = QPushButton("Agregar Curso")
        self.button_agregar_curso.clicked.connect(self.agregarCurso)
        self.layout.addWidget(self.button_agregar_curso)

        self.setLayout(self.layout)

    def agregarCurso(self):
        nombre_curso = self.nombre_curso_field.text()
        nombre_profesor = self.profesor_combo.currentText()

        if not nombre_curso or not nombre_profesor:
            QMessageBox.warning(self, "Error", "Todos los campos deben estar completos")
            return

        profesor_seleccionado = next((p for p in self.parent.profesores if p.getNombreCompleto() == nombre_profesor), None)
        if profesor_seleccionado:
            curso = Curso(nombre_curso, profesor_seleccionado)
            self.parent.cursos.append(curso)
            QMessageBox.information(self, "Éxito", "Curso agregado correctamente")
            self.nombre_curso_field.clear()
        else:
            QMessageBox.warning(self, "Error", "Profesor no encontrado")


class VentanaAgregarProfesor(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.parent = parent
        self.initUI()

    def initUI(self):
        self.layout = QVBoxLayout()

        self.label_nombre = QLabel("Nombre:")
        self.nombre_field = QLineEdit()
        self.layout.addWidget(self.label_nombre)
        self.layout.addWidget(self.nombre_field)

        self.label_apellido = QLabel("Apellido:")
        self.apellido_field = QLineEdit()
        self.layout.addWidget(self.label_apellido)
        self.layout.addWidget(self.apellido_field)

        self.button_agregar_profesor = QPushButton("Agregar Profesor")
        self.button_agregar_profesor.clicked.connect(self.agregarProfesor)
        self.layout.addWidget(self.button_agregar_profesor)

        self.setLayout(self.layout)

    def agregarProfesor(self):
        nombre = self.nombre_field.text()
        apellido = self.apellido_field.text()

        if not nombre or not apellido:
            QMessageBox.warning(self, "Error", "Todos los campos deben estar completos")
            return

        profesor = Profesor(nombre, apellido)
        self.parent.profesores.append(profesor)
        QMessageBox.information(self, "Éxito", "Profesor agregado correctamente")
        self.nombre_field.clear()
        self.apellido_field.clear()


class VentanaAgregarEstudiante(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.parent = parent
        self.initUI()

    def initUI(self):
        self.layout = QVBoxLayout()

        self.label_nombre = QLabel("Nombre:")
        self.nombre_field = QLineEdit()
        self.layout.addWidget(self.label_nombre)
        self.layout.addWidget(self.nombre_field)

        self.label_apellido = QLabel("Apellido:")
        self.apellido_field = QLineEdit()
        self.layout.addWidget(self.label_apellido)
        self.layout.addWidget(self.apellido_field)

        self.button_agregar_estudiante = QPushButton("Agregar Estudiante")
        self.button_agregar_estudiante.clicked.connect(self.agregarEstudiante)
        self.layout.addWidget(self.button_agregar_estudiante)

        self.setLayout(self.layout)

    def agregarEstudiante(self):
        nombre = self.nombre_field.text()
        apellido = self.apellido_field.text()

        if not nombre or not apellido:
            QMessageBox.warning(self, "Error", "Todos los campos deben estar completos")
            return

        estudiante = Estudiante(nombre, apellido)

        # Mostrar ventana para seleccionar cursos
        ventana_seleccion_curso = VentanaSeleccionCurso(estudiante, self.parent.cursos, self)
        ventana_seleccion_curso.exec_()


class VentanaSeleccionCurso(QDialog):

    def __init__(self, estudiante, cursos, parent=None):
        super().__init__(parent)
        self.estudiante = estudiante
        self.cursos = cursos
        self.initUI()

    def initUI(self):
        self.layout = QVBoxLayout()

        self.label_seleccion = QLabel("Seleccione el curso:")
        self.curso_combo = QComboBox()
        for curso in self.cursos:
            self.curso_combo.addItem(curso.nombre)
        self.layout.addWidget(self.label_seleccion)
        self.layout.addWidget(self.curso_combo)

        self.button_seleccionar_curso = QPushButton("Inscribir en Curso")
        self.button_seleccionar_curso.clicked.connect(self.inscribirEstudiante)
        self.layout.addWidget(self.button_seleccionar_curso)

        self.setLayout(self.layout)

    def inscribirEstudiante(self):
        nombre_curso = self.curso_combo.currentText()
        curso_seleccionado = next((c for c in self.cursos if c.nombre == nombre_curso), None)
        if curso_seleccionado:
            curso_seleccionado.agregarEstudiante(self.estudiante)
            QMessageBox.information(self, "Éxito", "Estudiante inscrito correctamente")
            self.accept()
        else:
            QMessageBox.warning(self, "Error", "Curso no encontrado")


class VentanaHorarios(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.parent = parent
        self.initUI()

    def initUI(self):
        self.layout = QVBoxLayout()

        self.text_area_horarios = QTextEdit()
        self.text_area_horarios.setReadOnly(True)
        self.layout.addWidget(self.text_area_horarios)

        self.setLayout(self.layout)

    def mostrarHorarios(self, horarios):
        self.text_area_horarios.clear()
        self.text_area_horarios.append(horarios)


if __name__ == "__main__":
    app = QApplication(sys.argv)
    ventana_principal = GestionDeCursos()
    ventana_principal.show()
    sys.exit(app.exec_())

File: Codigo Java/src/Gestiondecursos.java
package gestiondecursos;


import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.ArrayList;
import java.util.List;

public class Gestiondecursos {
    private List<Curso> cursos;
    private List<Profesor> profesores;

    public Gestiondecursos() {
        cursos = new ArrayList<>();
        profesores = new ArrayList<>();
        inicializarDatosPredeterminados();
    }

    // Método para inicializar datos predeterminados (profesores y cursos)
    private void inicializarDatosPredeterminados() {
        // Crear y agregar profesores predeterminados
        Profesor profesor1 = new Profesor("Juan", "Pérez");
        Profesor profesor2 = new Profesor("María", "González");
        Profesor profesor3 = new Profesor("Carlos", "Sánchez");
        Profesor profesor4 = new Profesor("Laura", "Martínez");
        Profesor profesor5 = new Profesor("Pedro", "López");

        profesores.add(profesor1);
        profesores.add(profesor2);
        profesores.add(profesor3);
        profesores.add(profesor4);
        profesores.add(profesor5);

        // Crear y agregar cursos predeterminados
        Curso curso1 = new Curso("Matemáticas", profesor1);
        Curso curso2 = new Curso("Historia", profesor2);
        Curso curso3 = new Curso("Programación", profesor3);
        Curso curso4 = new Curso("Biología", profesor4);
        Curso curso5 = new Curso("Literatura", profesor5);

        cursos.add(curso1);
        cursos.add(curso2);
        cursos.add(curso3);
        cursos.add(curso4);
        cursos.add(curso5);
    }

    // Método para abrir la ventana de agregar curso
    public void abrirVentanaAgregarCurso() {
        JFrame frame = new JFrame("Agregar Curso");
        frame.setSize(400, 300);
        frame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
        centrarVentanaEnPantalla(frame);

        JPanel panel = new JPanel();

        JLabel labelNombreCurso = new JLabel("Nombre del Curso:");
        JTextField nombreCursoField = new JTextField(20);
        panel.add(labelNombreCurso);
        panel.add(nombreCursoField);

        JLabel labelProfesor = new JLabel("Profesor:");
        JComboBox<String> profesorComboBox = new JComboBox<>();
        for (Profesor profesor : profesores) {
            profesorComboBox.addItem(profesor.getNombreCompleto());
        }
        panel.add(labelProfesor);
        panel.add(profesorComboBox);

            JButton buttonAgregarCurso = new JButton("Agregar Curso");
            buttonAgregarCurso.setPreferredSize(new Dimension(150, 30));        buttonAgregarCurso.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                String nombreCurso = nombreCursoField.getText();
                String nombreProfesor = (String) profesorComboBox.getSelectedItem();
                Profesor profesorSeleccionado = buscarProfesor(nombreProfesor);

                if (nombreCurso.isEmpty() || nombreProfesor.isEmpty() || profesorSeleccionado == null) {
                    JOptionPane.showMessageDialog(frame, "Todos los campos deben estar completos", "Error", JOptionPane.ERROR_MESSAGE);
                    return;
                }

                Curso curso = new Curso(nombreCurso, profesorSeleccionado);
                cursos.add(curso);

                JOptionPane.showMessageDialog(frame, "Curso agregado correctamente", "Éxito", JOptionPane.INFORMATION_MESSAGE);

                // Limpia los campos después de agregar el curso
                nombreCursoField.setText("");
            }
        });
        panel.add(buttonAgregarCurso);

        frame.add(panel);
        frame.setVisible(true);
    }

    // Método para abrir la ventana de agregar profesor
    public void abrirVentanaAgregarProfesor() {
        JFrame frame = new JFrame("Agregar Profesor");
        frame.setSize(400, 200);
        frame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
        centrarVentanaEnPantalla(frame);

        JPanel panel = new JPanel();

        JLabel labelNombre = new JLabel("Nombre:");
        JTextField nombreField = new JTextField(20);
        panel.add(labelNombre);
        panel.add(nombreField);

        JLabel labelApellido = new JLabel("Apellido:");
        JTextField apellidoField = new JTextField(20);
        panel.add(labelApellido);
        panel.add(apellidoField);

        JButton buttonAgregarProfesor = new JButton("Agregar Profesor");
        buttonAgregarProfesor.setPreferredSize(new Dimension(150, 30));
        buttonAgregarProfesor.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                String nombre = nombreField.getText();
                String apellido = apellidoField.getText();

                if (nombre.isEmpty() || apellido.isEmpty()) {
                    JOptionPane.showMessageDialog(frame, "Todos los campos deben estar completos", "Error", JOptionPane.ERROR_MESSAGE);
                    return;
                }

                Profesor profesor = new Profesor(nombre, apellido);
                profesores.add(profesor);

                JOptionPane.showMessageDialog(frame, "Profesor agregado correctamente", "Éxito", JOptionPane.INFORMATION_MESSAGE);

                // Limpia los campos después de agregar el profesor
                nombreField.setText("");
                apellidoField.setText("");
            }
        });
        panel.add(buttonAgregarProfesor);

        frame.add(panel);
        frame.setVisible(true);
    }

    // Método para abrir la ventana de agregar estudiante
    public void abrirVentanaAgregarEstudiante() {
        JFrame frame = new JFrame("Agregar Estudiante");
        frame.setSize(400, 200);
        frame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
        centrarVentanaEnPantalla(frame);

        JPanel panel = new JPanel();

        JLabel labelNombre = new JLabel("Nombre:");
        JTextField nombreField = new JTextField(20);
        panel.add(labelNombre);
        panel.add(nombreField);

        JLabel labelApellido = new JLabel("Apellido:");
        JTextField apellidoField = new JTextField(20);
        panel.add(labelApellido);
        panel.add(apellidoField);

        JButton buttonAgregarEstudiante = new JButton("Agregar Estudiante");
        buttonAgregarEstudiante.setPreferredSize(new Dimension(150, 30));        buttonAgregarEstudiante.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                String nombre = nombreField.getText();
                String apellido = apellidoField.getText();

                if (nombre.isEmpty() || apellido.isEmpty()) {
                    JOptionPane.showMessageDialog(frame, "Todos los campos deben estar completos", "Error", JOptionPane.ERROR_MESSAGE);
                    return;
                }

                Estudiante estudiante = new Estudiante(nombre, apellido);

                // Mostrar ventana para seleccionar cursos
                JFrame frameSeleccionCurso = new JFrame("Seleccionar Curso");
                frameSeleccionCurso.setSize(400, 300);
                frameSeleccionCurso.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
                centrarVentanaEnPantalla(frameSeleccionCurso);

                JPanel panelSeleccion = new JPanel();
                JLabel labelSeleccion = new JLabel("Seleccione el curso:");
                JComboBox<String> cursoComboBox = new JComboBox<>();
                for (Curso curso : cursos) {
                    cursoComboBox.addItem(curso.getNombre());
                }
                panelSeleccion.add(labelSeleccion);
                panelSeleccion.add(cursoComboBox);

                JButton buttonSeleccionarCurso = new JButton("Inscribir en Curso");
                buttonSeleccionarCurso.addActionListener(new ActionListener() {
                    @Override
                    public void actionPerformed(ActionEvent e) {
                        String nombreCurso = (String) cursoComboBox.getSelectedItem();
                        Curso cursoSeleccionado = buscarCurso(nombreCurso);

                        if (cursoSeleccionado != null) {
                            cursoSeleccionado.agregarEstudiante(estudiante);
                            JOptionPane.showMessageDialog(frameSeleccionCurso, "Estudiante inscrito correctamente", "Éxito", JOptionPane.INFORMATION_MESSAGE);
                            frameSeleccionCurso.dispose(); // Cierra la ventana de selección de curso
                            frame.dispose(); // Cierra la ventana de agregar estudiante
                        } else {
                            JOptionPane.showMessageDialog(frameSeleccionCurso, "Curso no encontrado", "Error", JOptionPane.ERROR_MESSAGE);
                        }
                    }
                });
                panelSeleccion.add(buttonSeleccionarCurso);

                frameSeleccionCurso.add(panelSeleccion);
                frameSeleccionCurso.setVisible(true);
            }
        });
        panel.add(buttonAgregarEstudiante);

        frame.add(panel);
        frame.setVisible(true);
    }

    // Método para abrir la ventana de horarios
    public void abrirVentanaHorarios() {
        JFrame frame = new JFrame("Horarios de Cursos");
        frame.setSize(600, 400);
        frame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
        centrarVentanaEnPantalla(frame);

        JPanel panel = new JPanel();

        JTextArea textAreaHorarios = new JTextArea(20, 40);
        textAreaHorarios.setEditable(false);
        panel.add(textAreaHorarios);

        // Mostrar información de horarios de cada curso
        StringBuilder horarios = new StringBuilder();
        for (Curso curso : cursos) {
            horarios.append("Curso: ").append(curso.getNombre()).append("\n");
            horarios.append("Profesor: ").append(curso.getProfesor().getNombreCompleto()).append("\n");
            horarios.append("Estudiantes:\n");
            for (Estudiante estudiante : curso.getEstudiantes()) {
                horarios.append("- ").append(estudiante.getNombreCompleto()).append("\n");
            }
            horarios.append("\n");
        }
        textAreaHorarios.setText(horarios.toString());

        frame.add(panel);
        frame.setVisible(true);
    }

        // Método para centrar una ventana en la pantalla
    private void centrarVentanaEnPantalla(JFrame frame) {
        Dimension dim = Toolkit.getDefaultToolkit().getScreenSize();
        int w = frame.getSize().width;
        int h = frame.getSize().height;
        int x = (dim.width - w) / 2;
        int y = (dim.height - h) / 2;
        frame.setLocation(x, y);
    }

    // Método para buscar un profesor por su nombre completo
    private Profesor buscarProfesor(String nombreCompleto) {
        for (Profesor profesor : profesores) {
            if (profesor.getNombreCompleto().equals(nombreCompleto)) {
                return profesor;
            }
        }
        return null;
    }

    // Método para buscar un curso por su nombre
    private Curso buscarCurso(String nombreCurso) {
        for (Curso curso : cursos) {
            if (curso.getNombre().equals(nombreCurso)) {
                return curso;
            }
        }
        return null;
    }

    // Clase principal que contiene el método main
    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> {
            Gestiondecursos gestiondecursos = new Gestiondecursos();
            gestiondecursos.crearInterfazGrafica();
        });
    }

    // Método para crear la interfaz gráfica principal con un menú organizado
    public void crearInterfazGrafica() {
        JFrame frame = new JFrame("Gestión de Cursos");
        frame.setSize(400, 300);
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        centrarVentanaEnPantalla(frame);

        JPanel panel = new JPanel(new GridBagLayout());
        GridBagConstraints gbc = new GridBagConstraints();
        gbc.fill = GridBagConstraints.HORIZONTAL;
        gbc.gridx = 0;
        gbc.gridy = 0;
        gbc.weightx = 1.0;
        gbc.weighty = 1.0;

        // Crear botones para las acciones principales
        JButton buttonAgregarCurso = new JButton("Agregar Curso");
        JButton buttonAgregarProfesor = new JButton("Agregar Profesor");
        JButton buttonAgregarEstudiante = new JButton("Agregar Estudiante");
        JButton buttonVerHorarios = new JButton("Ver Horarios");
        buttonVerHorarios.setPreferredSize(new Dimension(150, 30));
        // Configurar listeners para cada botón
        buttonAgregarCurso.addActionListener(e -> abrirVentanaAgregarCurso());
        buttonAgregarProfesor.addActionListener(e -> abrirVentanaAgregarProfesor());
        buttonAgregarEstudiante.addActionListener(e -> abrirVentanaAgregarEstudiante());
        buttonVerHorarios.addActionListener(e -> abrirVentanaHorarios());

        // Añadir botones al panel
        panel.add(buttonAgregarCurso, gbc);
        gbc.gridy++;
        panel.add(buttonAgregarProfesor, gbc);
        gbc.gridy++;
        panel.add(buttonAgregarEstudiante, gbc);
        gbc.gridy++;
        panel.add(buttonVerHorarios, gbc);

        // Agregar el panel al centro del frame
        frame.add(panel, BorderLayout.CENTER);

        // Hacer visible el frame
        frame.setVisible(true);
    }
}
